File: templates/python/v1/visitor/nodes.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import List, Protocol


class NodeVisitor(Protocol):
    def visit_text(self, node: "TextNode") -> str:
        ...

    def visit_element(self, node: "ElementNode") -> str:
        ...


class Node(Protocol):
    def accept(self, visitor: NodeVisitor) -> str:
        ...


@dataclass(slots=True)
class TextNode:
    text: str

    def accept(self, visitor: NodeVisitor) -> str:
        return visitor.visit_text(self)


@dataclass(slots=True)
class ElementNode:
    tag: str
    children: List[Node] = field(default_factory=list)

    def add(self, child: Node) -> "ElementNode":
        self.children.append(child)
        return self

    def accept(self, visitor: NodeVisitor) -> str:
        return visitor.visit_element(self)


@dataclass(slots=True)
class RenderVisitor:
    def visit_text(self, node: TextNode) -> str:
        return node.text

    def visit_element(self, node: ElementNode) -> str:
        inner = "".join(child.accept(self) for child in node.children)
        return f"<{node.tag}>{inner}</{node.tag}>"

File: templates/javascript/v1/visitor/nodes.js
export class TextNode {
  constructor(text) {
    this.text = text;
  }

  accept(visitor) {
    return visitor.visitText(this);
  }
}

export class ElementNode {
  constructor(tag) {
    this.tag = tag;
    this.children = [];
  }

  add(child) {
    this.children.push(child);
    return this;
  }

  accept(visitor) {
    return visitor.visitElement(this);
  }
}

export class RenderVisitor {
  visitText(textNode) {
    return textNode.text;
  }

  visitElement(elementNode) {
    const inner = elementNode.children.map((c) => c.accept(this)).join("");
    return `<${elementNode.tag}>${inner}</${elementNode.tag}>`;
  }
}
